src/main/java/net/minecraftforge/common/TierSortingRegistry.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.graph.ElementOrder;
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.MutableGraph;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.tags.Tag;
+import net.minecraft.world.item.Tier;
+import net.minecraft.world.item.Tiers;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraftforge.fml.loading.toposort.TopologicalSort;
+
+import javax.annotation.Nullable;
+import java.util.*;
+
+public class TierSortingRegistry
+{
+    public static synchronized void registerTier(Tier tier, ResourceLocation name, List<Object> after, List<Object> before)
+    {
+        if (tiers.containsKey(name))
+            throw new IllegalStateException("Duplicate tier name " + name);
+
+        processTier(tier, name, after, before);
+    }
+
+    public static List<Tier> getSortedTiers()
+    {
+        return sortedTiers;
+    }
+
+    @Nullable
+    public static Tier byName(ResourceLocation name)
+    {
+        return tiers.get(name);
+    }
+
+    @Nullable
+    public static ResourceLocation getName(Tier tier)
+    {
+        return tiers.inverse().get(tier);
+    }
+
+    public static boolean isCorrectTierForDrops(Tier tier, BlockState state)
+    {
+        for (int x = sortedTiers.indexOf(tier) + 1; x < sortedTiers.size(); x++) {
+            if (state.is(sortedTiers.get(x).getTag()))
+                return false;
+        }
+        return true;
+    }
+
+    // ===================== PRIVATE INTERNAL STUFFS BELOW THIS LINE ====================
+
+    private static ResourceLocation processTier(Tier tier, ResourceLocation name, List<Object> afters, List<Object> befores)
+    {
+        tiers.put(name, tier);
+        for(Object after : afters)
+        {
+            if (after instanceof String s)
+                edges.put(new ResourceLocation(s), name);
+            else if (after instanceof ResourceLocation rl)
+                edges.put(rl, name);
+            else if (after instanceof Tiers t)
```

Comment:
Why is this check against `Tiers` rather than `Tier`?

Reply:
I wrote that logic before the redesign, when only vanilla Tiers objects were supported by instance. This slipped through. :P